test/model/message_test.dart

```diff
@@ -123,6 +127,273 @@ void main() {
+      store.editMessage(messageId: message.id, content: 'new content');
+      checkNotifiedOnce();
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+      checkNotifiedOnce();
+    }));
+
+    test('request fails; take failed edit', () => awaitFakeAsync((async) async {
+      await prepareEditMessage();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      checkNotifiedOnce();
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+      checkNotifiedOnce();
+
+      check(store.takeFailedMessageEdit(message.id)).equals('new content');
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      checkNotifiedOnce();
+    }));
+
+    test('takeFailedMessageEdit throws StateError when nothing to take', () => awaitFakeAsync((async) async {
+      await prepareEditMessage();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      check(() => store.takeFailedMessageEdit(message.id)).throws<StateError>();
+    }));
+
+    test('editMessage throws StateError if editMessage already in progress for same message', () => awaitFakeAsync((async) async {
+      await prepareEditMessage();
+
+      connection.prepare(
+        json: UpdateMessageResult().toJson(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      async.elapse(Duration(milliseconds: 500));
+      check(connection.takeRequests()).length.equals(1);
+      checkNotifiedOnce();
+
+      // A function marked async needs an async check,
+      // but its return type is void, so, cast the return value to Future
+      final future = store.editMessage(messageId: message.id, content: 'newer content') as Future;
+      await check(future).throws<StateError>();
+      check(connection.takeRequests()).isEmpty();
+    }));
+
+    test('event arrives, then request fails', () => awaitFakeAsync((async) async {
+      // This can happen with network issues.
+
+      await prepareEditMessage();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(
+        httpException: const SocketException('failed'), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      checkNotifiedOnce();
+
+      async.elapse(Duration(milliseconds: 500));
+      await store.handleEvent(eg.updateMessageEditEvent(message));
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      checkNotifiedOnce();
+
+      async.elapse(Duration(milliseconds: 500));
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      checkNotNotified();
```

Comment:
This version is slightly delicate, in that these two checks are only effective given that the `elapse` calls add up to at least the delay on the error response. The test is correct, but vulnerable to being defused in a not-entirely-obvious way by a future edit to those numbers.

How about:
```suggestion
      async.flushTimers();
      check(store.getEditMessageErrorStatus(message.id)).isNull();
      checkNotNotified();
```

(It's basically OK as is, though; if there weren't such an easy way to make it more robust in that direction, I wouldn't want to attempt anything complex to guard against that sort of edit.)